### DISCUSSION TITLE
Avoiding node_modules in final build

### DISCUSSION THREAD

**Discussion by RicardoGaefke** [jaredpalmer/razzle#1781] [Q&A] [answered]
## ❓Question

First of all, thank you very much for your amazing work. I'm using razzle within an ASP.NET Core 5 (Jering.Javascript.NodeJS for ssr) app and it works pretty fine but there is something that I'd like to improve. Using Docker I need to copy `node_modules` to my final layer which makes it really bigger. This is my dockerfile:

```
# clones the app
FROM alpine AS gitclone
ARG BRANCH
ARG GIT_CRED

RUN apk update && apk add --no-cache git

RUN mkdir /root/app/
RUN git clone -b ${BRANCH} https://${GIT_CRED}@dev.azure.com/myCompany/myRepo/_git/any ./app

# builds the app
FROM mcr.microsoft.com/dotnet/sdk:5.0-alpine3.12 AS build

RUN apk update && apk add --no-cache nodejs npm

COPY --from=gitclone /app ./app

WORKDIR /app/src/WebSite

RUN dotnet publish -c Release ./WebSite.csproj -r linux-musl-x64 -o ./publish

# runs the app
FROM mcr.microsoft.com/dotnet/aspnet:5.0.4-alpine3.12 as runtime

RUN apk update && apk add --no-cache nodejs npm

RUN export PATH=$PATH:/usr/bin/node:/usr/share/doc/nodejs

ENV ASPNETCORE_URLS=http://+:5000

COPY --from=build /app/src/WebSite/publish ./
COPY --from=build /app/src/WebSite/ClientApp/node_modules ./node_modules
COPY --from=build /app/src/WebSite/ClientApp/build ./ClientApp/build

ENTRYPOINT ["./WebSite"]
```

The app wasn't working before copying the `node_modules` and I found this in your docs:

![image](https://user-images.githubusercontent.com/32141532/144775544-8813c20a-1647-40ba-a34b-7a52cb731c47.png)

Is there a way to build server bundle to avoid copying `node_modules`?

Thank you very much.

**Comment by fivethreeo** [comment, marked as answer]
You can do it like this https://github.com/fivethreeo/razzle-with-docker-nginx/blob/master/Dockerfile

or bundle everything with buildType iso-serverless in razzle options

**Reply by fivethreeo** [reply]
also, use buildkit for faster docker builds

**Reply by RicardoGaefke** [reply]
Thank you very much... `buildType: 'iso-serverless'` worked pretty fine. I'll take a look into buildkit.

**Reply by RicardoGaefke** [reply]
Sorry, it was my mistake. Using buildType: 'iso-serverless' didn't work for me and this is the message:

![image](https://user-images.githubusercontent.com/32141532/144902967-1ef1809d-73c8-4573-aa6e-18163f8095a4.png)

I'll keep copying the node_modules for now and I'll look for something to improve image size (without node_modules is half size).

**Reply by fivethreeo** [reply]
I think this may have to do with loadable-components and react being a lazy require. Think you may need webpack contextplugin to make sure react is required directly on serverside. In react 18 we can stop using loadable-components.

**Comment by fivethreeo** [comment]
If using express

https://github.com/jaredpalmer/razzle/tree/master/examples/basic-serverless

**Comment by fivethreeo** [comment]
Make sure react is imported in the server entrypoint so it is not dynamically imported

**Reply by fivethreeo** [reply]
Use https://github.com/jaredpalmer/razzle/tree/master/examples/with-development-build , to inspect what server.js does by reading built server.js source

**Comment by dmitryshelomanov** [comment]
@fivethreeo  not working for me

too many errors like this

```
./node_modules/react-router-dom/server.mjs 72:18-37
Can't import the named export 'createElement' from non EcmaScript module (only default export is available)
```

how to build server with node modules ?

**Comment by fivethreeo** [comment]
https://razzlejs.org/docs/customization#transpilation-of-external-modules
